File: apps/custom-harness/src/tui-bubbletea/protocol.ts
export const TUI_PROTOCOL_VERSION = 1 as const;
export const TUI_PROTOCOL_FRAME_LIMIT = 1 << 20;

export interface TuiCapabilitySnapshot {
  readonly id: string;
  readonly reason: string;
  readonly state:
    | "catalogued"
    | "scaffolded"
    | "available"
    | "qualified"
    | "unavailable";
}

export interface TuiCatalogSnapshot {
  readonly commands: readonly {
    readonly description: string;
    readonly name: string;
    readonly status: "active" | "compatibility-deprecated";
  }[];
  readonly digest: string;
  readonly pluginIds: readonly string[];
  readonly toolNames: readonly string[];
  readonly workflowNames: readonly string[];
}

export interface TuiHostSnapshot {
  readonly actorId: string;
  readonly capabilities: readonly TuiCapabilitySnapshot[];
  readonly catalog: TuiCatalogSnapshot;
  readonly effort: string;
  readonly error: string;
  readonly inspectorText: string;
  readonly messages: readonly {
    readonly role: "assistant" | "user";
    readonly sequence: number;
    readonly text: string;
  }[];
  readonly modelId: string;
  readonly profile: string;
  readonly status: "idle" | "working";
  readonly streamingText: string;
  readonly submittedText: string;
  readonly threadId: string;
  readonly threadTitle: string;
  readonly workingDirectory: string;
}

export type TuiClientMessage =
  | { readonly type: "client.hello"; readonly payload: { readonly nonce: string } }
  | {
      readonly type: "client.turn.submit";
      readonly payload: { readonly text: string };
    }
  | { readonly type: "client.quit"; readonly payload: Readonly<Record<string, never>> };

export interface TuiHostMessage {
  readonly payload: TuiHostSnapshot | { readonly code: string };
  readonly type: "host.error" | "host.snapshot";
}

interface WireEnvelope {
  readonly payload: unknown;
  readonly type: string;
  readonly version: number;
}

const record = (value: unknown, error: string): Record<string, unknown> => {
  if (!value || typeof value !== "object" || Array.isArray(value))
    throw new Error(error);
  return value as Record<string, unknown>;
};

const exactKeys = (
  value: Record<string, unknown>,
  keys: readonly string[],
  error: string,
): void => {
  const actual = Object.keys(value).sort();
  const expected = [...keys].sort();
  if (
    actual.length !== expected.length ||
    actual.some((key, index) => key !== expected[index])
  )
    throw new Error(error);
};

const decodePayload = (
  envelope: WireEnvelope,
): TuiClientMessage["payload"] => {
  const payload = record(envelope.payload, "TUI_PROTOCOL_PAYLOAD_INVALID");
  if (envelope.type === "client.quit") {
    exactKeys(payload, [], "TUI_PROTOCOL_PAYLOAD_INVALID");
    return Object.freeze({});
  }
  if (envelope.type === "client.hello") {
    exactKeys(payload, ["nonce"], "TUI_PROTOCOL_PAYLOAD_INVALID");
    if (typeof payload.nonce !== "string" || !/^[a-f0-9]{64}$/u.test(payload.nonce))
      throw new Error("TUI_PROTOCOL_NONCE_INVALID");
    return Object.freeze({ nonce: payload.nonce });
  }
  if (envelope.type === "client.turn.submit") {
    exactKeys(payload, ["text"], "TUI_PROTOCOL_PAYLOAD_INVALID");
    if (
      typeof payload.text !== "string" ||
      !payload.text.trim() ||
      Buffer.byteLength(payload.text) > 64 * 1024
    )
      throw new Error("TUI_PROTOCOL_TURN_INVALID");
    return Object.freeze({ text: payload.text });
  }
  throw new Error("TUI_PROTOCOL_MESSAGE_UNSUPPORTED");
};

export const decodeTuiClientMessage = (value: unknown): TuiClientMessage => {
  const envelope = record(value, "TUI_PROTOCOL_FRAME_INVALID");
  exactKeys(envelope, ["payload", "type", "version"], "TUI_PROTOCOL_FRAME_INVALID");
  if (envelope.version !== TUI_PROTOCOL_VERSION || typeof envelope.type !== "string")
    throw new Error("TUI_PROTOCOL_VERSION_UNSUPPORTED");
  const wire: WireEnvelope = {
    payload: envelope.payload,
    type: envelope.type,
    version: envelope.version,
  };
  return Object.freeze({
    payload: decodePayload(wire),
    type: wire.type,
  }) as TuiClientMessage;
};

export const encodeTuiHostMessage = (message: TuiHostMessage): string => {
  const frame = `${JSON.stringify({
    payload: message.payload,
    type: message.type,
    version: TUI_PROTOCOL_VERSION,
  })}\n`;
  if (Buffer.byteLength(frame) > TUI_PROTOCOL_FRAME_LIMIT)
    throw new Error("TUI_PROTOCOL_FRAME_TOO_LARGE");
  return frame;
};

File: apps/custom-harness/native/tui/internal/protocol/protocol.go
package protocol

import (
	"bufio"
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"sync"
)

const (
	Version       = 1
	MaximumFrame  = 1 << 20
	TypeHello     = "client.hello"
	TypeSnapshot  = "host.snapshot"
	TypeTurn      = "client.turn.submit"
	TypeQuit      = "client.quit"
	TypeHostError = "host.error"
)

type Envelope struct {
	Version int             `json:"version"`
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload"`
}

type Hello struct {
	Nonce string `json:"nonce"`
}

type TurnSubmit struct {
	Text string `json:"text"`
}

type HostError struct {
	Code string `json:"code"`
}

type Command struct {
	Description string `json:"description"`
	Name        string `json:"name"`
	Status      string `json:"status"`
}

type Capability struct {
	ID     string `json:"id"`
	Reason string `json:"reason"`
	State  string `json:"state"`
}

type Catalog struct {
	Digest        string    `json:"digest"`
	PluginIDs     []string  `json:"pluginIds"`
	Commands      []Command `json:"commands"`
	ToolNames     []string  `json:"toolNames"`
	WorkflowNames []string  `json:"workflowNames"`
}

type Message struct {
	Role     string `json:"role"`
	Sequence int    `json:"sequence"`
	Text     string `json:"text"`
}

type Snapshot struct {
	ActorID          string       `json:"actorId"`
	Capabilities     []Capability `json:"capabilities"`
	Catalog          Catalog      `json:"catalog"`
	Effort           string       `json:"effort"`
	Error            string       `json:"error"`
	Messages         []Message    `json:"messages"`
	ModelID          string       `json:"modelId"`
	Profile          string       `json:"profile"`
	Status           string       `json:"status"`
	StreamingText    string       `json:"streamingText"`
	SubmittedText    string       `json:"submittedText"`
	ThreadID         string       `json:"threadId"`
	ThreadTitle      string       `json:"threadTitle"`
	WorkingDirectory string       `json:"workingDirectory"`
}

type Codec struct {
	scanner *bufio.Scanner
	writer  *bufio.Writer
	mu      sync.Mutex
}

func NewCodec(reader io.Reader, writer io.Writer) *Codec {
	scanner := bufio.NewScanner(reader)
	scanner.Buffer(make([]byte, 4096), MaximumFrame)
	return &Codec{scanner: scanner, writer: bufio.NewWriter(writer)}
}

func (codec *Codec) Receive() (Envelope, error) {
	if !codec.scanner.Scan() {
		if err := codec.scanner.Err(); err != nil {
			return Envelope{}, fmt.Errorf("PROTOCOL_READ_FAILED: %w", err)
		}
		return Envelope{}, io.EOF
	}
	return decodeEnvelope(codec.scanner.Bytes())
}

func (codec *Codec) Send(messageType string, payload any) error {
	if messageType == "" {
		return errors.New("PROTOCOL_TYPE_REQUIRED")
	}
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("PROTOCOL_PAYLOAD_INVALID: %w", err)
	}
	frame, err := json.Marshal(Envelope{
		Version: Version,
		Type:    messageType,
		Payload: payloadBytes,
	})
	if err != nil {
		return fmt.Errorf("PROTOCOL_FRAME_INVALID: %w", err)
	}
	if len(frame) > MaximumFrame {
		return errors.New("PROTOCOL_FRAME_TOO_LARGE")
	}
	codec.mu.Lock()
	defer codec.mu.Unlock()
	if _, err := codec.writer.Write(append(frame, '\n')); err != nil {
		return fmt.Errorf("PROTOCOL_WRITE_FAILED: %w", err)
	}
	if err := codec.writer.Flush(); err != nil {
		return fmt.Errorf("PROTOCOL_FLUSH_FAILED: %w", err)
	}
	return nil
}

func DecodePayload[T any](envelope Envelope) (T, error) {
	var result T
	decoder := json.NewDecoder(bytes.NewReader(envelope.Payload))
	decoder.DisallowUnknownFields()
	if err := decoder.Decode(&result); err != nil {
		return result, fmt.Errorf("PROTOCOL_PAYLOAD_INVALID: %w", err)
	}
	if decoder.Decode(&struct{}{}) != io.EOF {
		return result, errors.New("PROTOCOL_PAYLOAD_TRAILING_DATA")
	}
	return result, nil
}

func decodeEnvelope(frame []byte) (Envelope, error) {
	if len(frame) == 0 || len(frame) > MaximumFrame {
		return Envelope{}, errors.New("PROTOCOL_FRAME_INVALID")
	}
	decoder := json.NewDecoder(bytes.NewReader(frame))
	decoder.DisallowUnknownFields()
	var envelope Envelope
	if err := decoder.Decode(&envelope); err != nil {
		return Envelope{}, fmt.Errorf("PROTOCOL_FRAME_INVALID: %w", err)
	}
	if decoder.Decode(&struct{}{}) != io.EOF {
		return Envelope{}, errors.New("PROTOCOL_FRAME_TRAILING_DATA")
	}
	if envelope.Version != Version {
		return Envelope{}, errors.New("PROTOCOL_VERSION_UNSUPPORTED")
	}
	if envelope.Type == "" || len(envelope.Payload) == 0 {
		return Envelope{}, errors.New("PROTOCOL_ENVELOPE_INCOMPLETE")
	}
	return envelope, nil
}
